addons/web/static/src/core/template_inheritance.js

```diff
@@ -0,0 +1,276 @@
+            if (textNode.previousSibling.nodeType === Node.TEXT_NODE) {
+                mergeTextNodes(textNode.previousSibling, textNode);
+            }
+        }
+    }
+}
+
+function getAncestor(node) {
+    if (node.parentElement === null) {
+        return node;
+    }
+    return getAncestor(node.parentElement);
+}
+
+const HASCLASS_REGEXP = /hasclass\(([^)]*)\)/;
+function replaceEveryHasclass(str) {
+    const result = HASCLASS_REGEXP.exec(str);
+    if (!result) {
+        return str;
+    }
+    const index = result.index;
+    const match = result[0];
+    const group = result[1];
+    const params = group.split(",");
+    const replacement = params
+        .map((param) => {
+            const p = param.trim();
+            if (p.length == 2) {
+                return `false`;
+            }
+            const pstemp = p.split("");
+            pstemp.splice(pstemp.length - 1, 0, " ");
+            const ps = pstemp.join("");
+            const sptemp = p.split("");
+            sptemp.splice(1, 0, " ");
+            const sp = sptemp.join("");
+            const spstemp = p.split("");
+            spstemp.splice(1, 0, " ");
+            spstemp.splice(spstemp.length - 1, 0, " ");
+            const sps = spstemp.join("");
+            return `(@class=${p} or starts-with(@class,${ps}) or contains(@class,${sps}) or (substring(@class,string-length(@class)-string-length(${sp})+1)=${sp}))`;
+        })
+        .join(" and ");
+    return (
+        str.slice(0, index) + replacement + replaceEveryHasclass(str.slice(index + match.length))
+    );
+}
+
+function getXpath(node) {
+    const xpath = node.getAttribute("expr");
+    // hasclass does not exist in XPath 1.0 but is a custom function defined browser side (see _hasclass) usable in lxml.
+    // Here we have to replace it by a complex condition (which is not nice).
+    return replaceEveryHasclass(xpath);
+}
+
+function getNode(source, node) {
+    const ancestor = getAncestor(source);
+    const doc = new Document();
+    doc.appendChild(ancestor);
+    if (node.tagName === "xpath") {
+        const xpath = getXpath(node);
+        const result = doc.evaluate(xpath, ancestor, null, XPathResult.FIRST_ORDERED_NODE_TYPE);
+        return result.singleNodeValue;
+    }
+    for (const elem of ancestor.querySelectorAll(node.tagName)) {
+        if (
+            [...node.attributes].every(
+                ({ name, value }) => name === "position" || elem.getAttribute(name) === value
+            )
+        ) {
+            return elem;
+        }
+    }
+    return null;
+}
+
+function getNodes(target, node) {
+    const nodes = [];
+    for (const childNode of node.childNodes) {
+        if (childNode.tagName === "xpath" && childNode.getAttribute?.("position") === "move") {
+            const node = getNode(target, childNode);
+            removeNode(node);
+            nodes.push(node);
+        } else {
+            nodes.push(childNode);
+        }
```

Comment:
why?

Reply:
I don't think I understand what this function is supposed to do, and why in the case of an xpath it finds the targeted nodes and returns them but in case of an operation described without an xpath it returns the operation itself.

Reply:
Consider the following example
```xml
<xpath expr="./div[1]" position="append">
   <span>Hello</span>
   <xpath expr="./div[2]" position="move"/>
   World
</xpath>
```
With that operation
you have to create to new element in the first div (the span and the text element "World") and you have to move the second div to the first one.
The way we create the two elements is to move them from operation to the first div
The move is done by finding the second div in the root  then append it to the first div

Reply:
That clarifies it, thanks. I thought this was operating on nodes one level higher.